Add a method for configuring items and prices via a file.

ahaywardtvuk opened the pull request:
For proper display in the UI we need to be able to set product names and, if any are ever chosen, prices.

This adds a way to do that by giving the accounting service's ingester a configuration file containing them.

ahaywardtvuk committed:
Add a method for configuring items and prices via a file.

shahidraza-telespazio commented on accounting_service/db.py:
```diff
@@ -56,3 +57,29 @@ def drop_tables():
+        valid_from: "2025-01-01T00:00:00Z"
+        price: 12.34
+    """
+    config_obj = yaml.safe_load(config)
```
This could potentially return None?
I was thinking of `config_obj = yaml.safe_load(config) or {}`

ahaywardtvuk replied:
I've changed this to detect YAML errors and an inappropriate return type and to be a bit more explicit in saying that it's the configuration file that's the cause. I do want an exception if an invalid configuration file is given, though, as it's only us who supply these and we should be informed if we write an invalid one.

The only way it can return None is if the configuration file consists of the text `null`, which is a very unlikely error in itself.

shahidraza-telespazio commented on accounting_service/db.py:
```diff
@@ -56,3 +57,29 @@ def drop_tables():
+    config_obj = yaml.safe_load(config)
+
+    with Session(engine) as session:
+        for item in config_obj.get("items", []):
```
Do we need to add any exception handling or this needs to abort during error?

ahaywardtvuk replied:
It should abort. An invalid config file needs to be picked up by an admin and not ignored.

If this occurs in the production service then a rolling rollout should occur, so only the new replacement pods will fail to start and the old ones will keep running. That means the service won't break, but it'll still show an error in Kubernetes, in the logs and in ArgoCD.

ahaywardtvuk replied:
It should abort.

shahidraza-telespazio commented on accounting_service/models.py:
```diff
@@ -192,6 +210,64 @@ def find_prices(cls, session: Session, at: datetime) -> Result[tuple[Self, str]]
+            print(f"{latest_price=}")
+            if latest_price.valid_from.astimezone(timezone.utc) > valid_from:
+                raise ValueError(
+                    f"Attempt to add price {price["sku"]} where valid_from is earlier "
```
`"` are mized so probably we will have to use this.
```
raise ValueError(
    f'Attempt to add price {price["sku"]} where valid_from is earlier '
    f'than the latest existing price, {latest_price.valid_from}.'
)
```

ahaywardtvuk replied:
Using the same quote type within `{}` inside an f-string is allowed from Python 3.12 onwards, so this code works OK.

shahidraza-telespazio commented on accounting_service/models.py:
```diff
@@ -192,6 +210,64 @@ def find_prices(cls, session: Session, at: datetime) -> Result[tuple[Self, str]]
+        item_obj = BillingItem.find_billing_item(session, price["sku"])
+        if not item_obj:
+            logging.error("Failed to find item %s when configuring price", price["sku"])
+            raise ValueError(f"Attempt to add price for unknown SKU {price["sku"]}")
```
"  are mixed so probably we will have to use this. Could throw Bad f-string error?

shahidraza-telespazio commented on accounting_service/models.py:
```diff
@@ -192,6 +210,64 @@ def find_prices(cls, session: Session, at: datetime) -> Result[tuple[Self, str]]
+        )
+
+        if latest_price:
+            print(f"{latest_price=}")
```
This may need to be logged using logger?

ahaywardtvuk replied:
I've removed the print - there's no need to log this.

shahidraza-telespazio commented on tests/test_package.py:
```diff
@@ -1,5 +1,113 @@
+    assert bi.sku == test_sku
+
+    prices = (
+        db_session.execute(select(BillingItemPrice).where(BillingItemPrice.item == bi))
```
Do we need to handle the exception?

ahaywardtvuk replied:
Not in a test, no. We would want the test to fail.

shahidraza-telespazio commented on accounting_service/models.py:
```diff
@@ -192,6 +210,64 @@ def find_prices(cls, session: Session, at: datetime) -> Result[tuple[Self, str]]
+        valid_from = datetime.fromisoformat(price["valid_from"]).astimezone(timezone.utc)
+
+        existing_price = (
+            session.execute(
```
Do we need to handle the exception for all the `session.execute`

ahaywardtvuk replied:
Is there anything you think we should do in response to the exception that won't happen automatically anyway?

shahidraza-telespazio replied:
I can only think of wrapping each upsert in try/except, log the specific SKU, then continue (or aggregate errors to raise at the end).

shahidraza-telespazio commented on accounting_service/models.py:
```diff
@@ -192,6 +210,64 @@ def find_prices(cls, session: Session, at: datetime) -> Result[tuple[Self, str]]
+                select(cls)
+                .where(cls.item == item_obj)
+                .where(cls.valid_from == valid_from)
+                .order_by(cls.price == price["price"])
```
`cls.price == price["price"]`  is boolean expression so do we need the `order_by` ?

shahidraza-telespazio replied:
Probably this could go like this but I am not entirely sure?
```
existing = session.scalar(
        select(cls)
        .where(cls.item_id == item_obj.id)
        .where(cls.valid_from == valid_from)
        .where(cls.price == price["price"])
    )
```

ahaywardtvuk replied:
Ordering by `price in db == new price` was meant to make the code prefer an exact match on all fields over a match only on item or valid_from, but not to require it (though it needed a `.desc()` to work as intended). It definitely shouldn't ignore existing prices where the price itself doesn't match - the point here is to update the existing price to match the new configuration.

There shouldn't be multiple BillingItemPrices for the same item starting at the same time so this was defensive. I've now changed it to run an update statement and update all copies of the price which match the item and start time.

shahidraza-telespazio reviewed (changes requested):
Looks good. A few comments

ahaywardtvuk committed:
If we see multiple BillingItemPrices with the same start time and a new price
has been configured for that start them then update all of them, not just one.
This situation shouldn't happen so this is being defensive.

ahaywardtvuk committed:
Be more explicit when exceptions are raised due to an invalid config file that
it is the config file at fault.

ahaywardtvuk committed:
Python 3.13 is targetted and 3.12 works, but 3.11 is likely to fail so
write this into the pyproject.toml.

shahidraza-telespazio reviewed (approved):
Thanks
